src/Pirouette/SMT/Constraints.hs

```diff
@@ -137,14 +132,16 @@ conjunct c cs0 =
       meta ->
       TermMeta lang meta ->
       Maybe (ConstraintSet lang meta)
-    unifyMetaWith cs v u
-      -- We check whether v is equivalent to some other smaller metavariable
-      | Just v' <- M.lookup v (csMetaEq cs) = unifyMetaWith cs v' u
-      -- If not, check whether @v@ is actually defined to be a given term. If so, that
-      -- have to be unifiable with u:
-      | Just t <- M.lookup v (csAssignments cs) = unifyWith cs t u
-      -- Finally, we have a /new/ @v@
-      | otherwise = Just $ unifyNewMetaWith cs v u
+    unifyMetaWith cs v u =
+      -- FIXME: We are dropping the optimised union-find here, by using @snd@.
+      case snd $ UF.lookup v (csAssignments cs) of
+        -- @v@ is known to be a given term that has to be unfifiable with @u@
+        Just (Just t) -> unifyWith cs t u
+        -- @v@ is known but does not link to any value. FIXME: Therefore, it
+        -- looks like we are violating an invariant of `unifyNewMetaWith`.
+        Just Nothing -> Just $ unifyNewMetaWith cs v u
```

Comment:
Maybe we can just change the name of `unifyNewMetaWith`? Perhaps something like `unifyUnassignedMetaWith`?

Reply:
This and updating the comment would make sense, at least. But maybe I will get rid of `unify*MetaWith` altogether,
cf https://github.com/tweag/pirouette/pull/150#discussion_r969783049 and https://github.com/tweag/pirouette/pull/150#issuecomment-1246358175.

Reply:
Commit c92f454 inlines `unifyNewMetaWith` into `unifyMetaWith` because the code had become very short. It also adds a bunch of comments to make things clearer. I actually understood a lot in the process. In particular, why the trivial union and trivial insert worked. In fact, I also understood the whole `unifyWith` function better and why it never needs a non-trivial union or insert. A full blown union-find retrospectively sounds overkill.